## Here we have implemented different categorical encoding techniques on the titanic dataset

In [3]:
import pandas as pd
import numpy as np

In [59]:
df = pd.read_csv('Titanic.csv')
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [60]:
PClassDict =   {   1:'First',
                    2:'Second',
                     3:'Third',              
                }

In [61]:
## changing the pclass column into categorical
df['Pclass'] = df.Pclass.map(PClassDict)
df

PassengerId  Survived  Pclass  \
0              1         0   Third   
1              2         1   First   
2              3         1   Third   
3              4         1   First   
4              5         0   Third   
..           ...       ...     ...   
886          887         0  Second   
887          888         1   First   
888          889         0   Third   
889          890         1   First   
890          891         0   Third   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [52]:
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin            0
Embarked         2
dtype: int64

In [53]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass          object
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [6]:
df['Age'].isna().sum()

177

In [7]:
df_na = df.loc[df['Age'].isna()]
df_na

PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex  Age  SibSp  Parch      Ticket     Fare Cabin Embarked  
5      male  NaN      0      0      330877   8.4583   NaN        Q  
17     male  NaN      0      0      244373  13.0000   NaN        S  
19   female  NaN      0      0        2649   7.2250   NaN        C  
26     male  NaN      0      0        2631   7.2250   NaN        C  
28   female  NaN      0      0      330959   7.8792   NaN        Q  
..      ...  ...    ...    ...         ...      ...   ...      ...  
859    male  NaN      0      0        2629   7.2292   NaN        C  
863  female  NaN      8      2    CA. 2343  69.5500   NaN        S  
868    male  NaN      0      0      345777   9.5000   NaN        S  
878    male  NaN      0      0      349217   7.8958   NaN        S  
888  female  NaN      1      2  W./C. 6607  23.4500   NaN        S  

[177 rows x 12 columns]

In [10]:
df_na['Age'].fillna(df['Age'].mean(), inplace=True)
df_na

C:\Users\atulkumarrai\anaconda3\lib\site-packages\pandas\core\series.py:4517: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().fillna(


PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex        Age  SibSp  Parch      Ticket     Fare Cabin Embarked  
5      male  29.699118      0      0      330877   8.4583   NaN        Q  
17     male  29.699118      0      0      244373  13.0000   NaN        S  
19   female  29.699118      0      0        2649   7.2250   NaN        C  
26     male  29.699118      0      0        2631   7.2250   NaN        C  
28   female  29.699118      0      0      330959   7.8792   NaN        Q  
..      ...        ...    ...    ...         ...      ...   ...      ...  
859    male  29.699118      0      0        2629   7.2292   NaN        C  
863  female  29.699118      8      2    CA. 2343  69.5500   NaN        S  
868    male  29.699118      0      0      345777   9.5000   NaN        S  
878    male  29.699118      0      0      349217   7.8958   NaN        S  
888  female  29.699118      1      2  W./C. 6607  23.4500   NaN        S  

[177 rows x 12 columns]

In [24]:
df['Cabin'].mode()

0        B96 B98
1    C23 C25 C27
2             G6
dtype: object

In [39]:
## Let's drop the missing values rows
print(df.shape)
df1 = df.dropna(axis=0, how='any')
print(df1.shape)
df1

(891, 12)
(183, 12)


PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
10            11         1       3   
11            12         1       1   
..           ...       ...     ...   
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
..                                                 ...     ...   ...    ...   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
872                           Carlsson, Mr. Frans Olof    male  33.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch    Ticket     Fare        Cabin Embarked  
1        0  PC 17599  71.2833          C85        C  
3        0    113803  53.1000         C123        S  
6        0     17463  51.8625          E46        S  
10       1   PP 9549  16.7000           G6        S  
11       0    113783  26.5500         C103        S  
..     ...       ...      ...          ...      ...  
871      1     11751  52.5542          D35        S  
872      0       695   5.0000  B51 B53 B55        S  
879      1     11767  83.1583          C50        C  
887      0    112053  30.0000          B42        S  
889      0    111369  30.0000         C148        C  

[183 rows x 12 columns]

In [42]:
## We observe that dropping will lead to major data loss, hence fill the cabin NAN values with the mode values
df['Cabin'].fillna(df['Cabin'].mode()[0], inplace=True)
df.isna().sum()


PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin            0
Embarked         2
dtype: int64

### 1. Dummies, it is equivallent to one hot encoding

In [44]:
df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare    Cabin Embarked  
0      0         A/5 21171   7.2500  B96 B98        S  
1      0          PC 17599  71.2833      C85        C  
2      0  STON/O2. 3101282   7.9250       G6        S  
3      0            113803  53.1000     C123        S  
4      0            373450   8.0500  B96 B98        S

In [46]:
data_dummies = pd.get_dummies(df)
data_dummies

PassengerId  Survived  Pclass   Age  SibSp  Parch     Fare  \
0              1         0       3  22.0      1      0   7.2500   
1              2         1       1  38.0      1      0  71.2833   
2              3         1       3  26.0      0      0   7.9250   
3              4         1       1  35.0      1      0  53.1000   
4              5         0       3  35.0      0      0   8.0500   
..           ...       ...     ...   ...    ...    ...      ...   
886          887         0       2  27.0      0      0  13.0000   
887          888         1       1  19.0      0      0  30.0000   
888          889         0       3   NaN      1      2  23.4500   
889          890         1       1  26.0      0      0  30.0000   
890          891         0       3  32.0      0      0   7.7500   

     Name_Abbing, Mr. Anthony  Name_Abbott, Mr. Rossmore Edward  \
0                           0                                 0   
1                           0                                 0   
2                           0                                 0   
3                           0                                 0   
4                           0                                 0   
..                        ...                               ...   
886                         0                                 0   
887                         0                                 0   
888                         0                                 0   
889                         0                                 0   
890                         0                                 0   

     Name_Abbott, Mrs. Stanton (Rosa Hunt)  ...  Cabin_F G73  Cabin_F2  \
0                                        0  ...            0         0   
1                                        0  ...            0         0   
2                                        0  ...            0         0   
3                                        0  ...            0         0   
4                                        0  ...            0         0   
..                                     ...  ...          ...       ...   
886                                      0  ...            0         0   
887                                      0  ...            0         0   
888                                      0  ...            0         0   
889                                      0  ...            0         0   
890                                      0  ...            0         0   

     Cabin_F33  Cabin_F38  Cabin_F4  Cabin_G6  Cabin_T  Embarked_C  \
0            0          0         0         0        0           0   
1            0          0         0         0        0           1   
2            0          0         0         1        0           0   
3            0          0         0         0        0           0   
4            0          0         0         0        0           0   
..         ...        ...       ...       ...      ...         ...   
886          0          0         0         0        0           0   
887          0          0         0         0        0           0   
888          0          0         0         0        0           0   
889          0          0         0         0        0           1   
890          0          0         0         0        0           0   

     Embarked_Q  Embarked_S  
0             0           1  
1             0           0  
2             0           1  
3             0           1  
4             0           1  
..          ...         ...  
886           0           1  
887           0           1  
888           0           1  
889           0           0  
890           1           0  

[891 rows x 1731 columns]

#### The above has one hot encoded all the categorical features and hence we have got 1731 columns now instead of 12 columns in the starting. So we will only select few columns for the one hot encoding and check

In [47]:
gender_dummies = pd.get_dummies(df['Sex'])
gender_dummies

female  male
0         0     1
1         1     0
2         1     0
3         1     0
4         0     1
..      ...   ...
886       0     1
887       1     0
888       1     0
889       0     1
890       0     1

[891 rows x 2 columns]

In [54]:
Pclass_dummies = pd.get_dummies(df['Pclass'])
Pclass_dummies

First  Second  Third
0        0       0      1
1        1       0      0
2        0       0      1
3        1       0      0
4        0       0      1
..     ...     ...    ...
886      0       1      0
887      1       0      0
888      0       0      1
889      1       0      0
890      0       0      1

[891 rows x 3 columns]

In [48]:
cabin_dummies = pd.get_dummies(df['Cabin'])
cabin_dummies

A10  A14  A16  A19  A20  A23  A24  A26  A31  A32  ...  E8  F E69  F G63  \
0      0    0    0    0    0    0    0    0    0    0  ...   0      0      0   
1      0    0    0    0    0    0    0    0    0    0  ...   0      0      0   
2      0    0    0    0    0    0    0    0    0    0  ...   0      0      0   
3      0    0    0    0    0    0    0    0    0    0  ...   0      0      0   
4      0    0    0    0    0    0    0    0    0    0  ...   0      0      0   
..   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ..    ...    ...   
886    0    0    0    0    0    0    0    0    0    0  ...   0      0      0   
887    0    0    0    0    0    0    0    0    0    0  ...   0      0      0   
888    0    0    0    0    0    0    0    0    0    0  ...   0      0      0   
889    0    0    0    0    0    0    0    0    0    0  ...   0      0      0   
890    0    0    0    0    0    0    0    0    0    0  ...   0      0      0   

     F G73  F2  F33  F38  F4  G6  T  
0        0   0    0    0   0   0  0  
1        0   0    0    0   0   0  0  
2        0   0    0    0   0   1  0  
3        0   0    0    0   0   0  0  
4        0   0    0    0   0   0  0  
..     ...  ..  ...  ...  ..  .. ..  
886      0   0    0    0   0   0  0  
887      0   0    0    0   0   0  0  
888      0   0    0    0   0   0  0  
889      0   0    0    0   0   0  0  
890      0   0    0    0   0   0  0  

[891 rows x 147 columns]

### 2. Ordinal Encoding 

In [62]:
## Here we will create a dictionary and then map it to the categorical columns
PClassDict =   {   'First':1,
                    'Second':2,
                     'Third':3,              
                }
df['Ordninal_Pclass'] = df.Pclass.map(PClassDict)
df

PassengerId  Survived  Pclass  \
0              1         0   Third   
1              2         1   First   
2              3         1   Third   
3              4         1   First   
4              5         0   Third   
..           ...       ...     ...   
886          887         0  Second   
887          888         1   First   
888          889         0   Third   
889          890         1   First   
890          891         0   Third   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  Ordninal_Pclass  
0        0         A/5 21171   7.2500   NaN        S                3  
1        0          PC 17599  71.2833   C85        C                1  
2        0  STON/O2. 3101282   7.9250   NaN        S                3  
3        0            113803  53.1000  C123        S                1  
4        0            373450   8.0500   NaN        S                3  
..     ...               ...      ...   ...      ...              ...  
886      0            211536  13.0000   NaN        S                2  
887      0            112053  30.0000   B42        S                1  
888      2        W./C. 6607  23.4500   NaN        S                3  
889      0            111369  30.0000  C148        C                1  
890      0            370376   7.7500   NaN        Q                3  

[891 rows x 13 columns]

### 3. Count/Frequency Encoding

In [66]:
Pclass_dict = df['Pclass'].value_counts()
df['Frequency_Encoded_Pclass'] = df.Pclass.map(Pclass_dict)
df

PassengerId  Survived  Pclass  \
0              1         0   Third   
1              2         1   First   
2              3         1   Third   
3              4         1   First   
4              5         0   Third   
..           ...       ...     ...   
886          887         0  Second   
887          888         1   First   
888          889         0   Third   
889          890         1   First   
890          891         0   Third   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  Ordninal_Pclass  \
0        0         A/5 21171   7.2500   NaN        S                3   
1        0          PC 17599  71.2833   C85        C                1   
2        0  STON/O2. 3101282   7.9250   NaN        S                3   
3        0            113803  53.1000  C123        S                1   
4        0            373450   8.0500   NaN        S                3   
..     ...               ...      ...   ...      ...              ...   
886      0            211536  13.0000   NaN        S                2   
887      0            112053  30.0000   B42        S                1   
888      2        W./C. 6607  23.4500   NaN        S                3   
889      0            111369  30.0000  C148        C                1   
890      0            370376   7.7500   NaN        Q                3   

     Frequency_Encoded_Pclass  
0                         491  
1                         216  
2                         491  
3                         216  
4                         491  
..                        ...  
886                       184  
887                       216  
888                       491  
889                       216  
890                       491  

[891 rows x 14 columns]

In [67]:
Cabin_dict = df['Cabin'].value_counts()
df['Encoded_cabin'] = df.Cabin.map(Cabin_dict)
df

PassengerId  Survived  Pclass  \
0              1         0   Third   
1              2         1   First   
2              3         1   Third   
3              4         1   First   
4              5         0   Third   
..           ...       ...     ...   
886          887         0  Second   
887          888         1   First   
888          889         0   Third   
889          890         1   First   
890          891         0   Third   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  Ordninal_Pclass  \
0        0         A/5 21171   7.2500   NaN        S                3   
1        0          PC 17599  71.2833   C85        C                1   
2        0  STON/O2. 3101282   7.9250   NaN        S                3   
3        0            113803  53.1000  C123        S                1   
4        0            373450   8.0500   NaN        S                3   
..     ...               ...      ...   ...      ...              ...   
886      0            211536  13.0000   NaN        S                2   
887      0            112053  30.0000   B42        S                1   
888      2        W./C. 6607  23.4500   NaN        S                3   
889      0            111369  30.0000  C148        C                1   
890      0            370376   7.7500   NaN        Q                3   

     Frequency_Encoded_Pclass  Encoded_cabin  
0                         491            NaN  
1                         216            1.0  
2                         491            NaN  
3                         216            2.0  
4                         491            NaN  
..                        ...            ...  
886                       184            NaN  
887                       216            1.0  
888                       491            NaN  
889                       216            1.0  
890                       491            NaN  

[891 rows x 15 columns]

### 4. Target/Guided encoding

In [71]:
df['Cabin_first_letter'] = df['Cabin'].astype(str).str[0]   ## create a new column and store the first letter of the cabin values
df.head(6)

PassengerId  Survived Pclass  \
0            1         0  Third   
1            2         1  First   
2            3         1  Third   
3            4         1  First   
4            5         0  Third   
5            6         0  Third   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   

   Parch            Ticket     Fare Cabin Embarked  Ordninal_Pclass  \
0      0         A/5 21171   7.2500   NaN        S                3   
1      0          PC 17599  71.2833   C85        C                1   
2      0  STON/O2. 3101282   7.9250   NaN        S                3   
3      0            113803  53.1000  C123        S                1   
4      0            373450   8.0500   NaN        S                3   
5      0            330877   8.4583   NaN        Q                3   

   Frequency_Encoded_Pclass  Encoded_cabin Cabin_first_letter  
0                       491            NaN                  n  
1                       216            1.0                  C  
2                       491            NaN                  n  
3                       216            2.0                  C  
4                       491            NaN                  n  
5                       491            NaN                  n

In [92]:
Encoded_labels = df.groupby(['Cabin'])['Survived'].mean().sort_values().index
Encoded_labels

Index(['A10', 'B86', 'B94', 'C110', 'C111', 'C118', 'C124', 'C128', 'C30',
       'C46',
       ...
       'D10 D12', 'B50', 'C99', 'B41', 'C93', 'C92', 'B18', 'C90', 'D11',
       'C62 C64'],
      dtype='object', name='Cabin', length=147)

In [90]:
Encoded_labels_rank = {k:i for i,k in enumerate(Encoded_labels,0)}
Encoded_labels_rank

{'A10': 0,
 'A14': 1,
 'A16': 2,
 'A19': 3,
 'A20': 4,
 'A23': 5,
 'A24': 6,
 'A26': 7,
 'A31': 8,
 'A32': 9,
 'A34': 10,
 'A36': 11,
 'A5': 12,
 'A6': 13,
 'A7': 14,
 'B101': 15,
 'B102': 16,
 'B18': 17,
 'B19': 18,
 'B20': 19,
 'B22': 20,
 'B28': 21,
 'B3': 22,
 'B30': 23,
 'B35': 24,
 'B37': 25,
 'B38': 26,
 'B39': 27,
 'B4': 28,
 'B41': 29,
 'B42': 30,
 'B49': 31,
 'B5': 32,
 'B50': 33,
 'B51 B53 B55': 34,
 'B57 B59 B63 B66': 35,
 'B58 B60': 36,
 'B69': 37,
 'B71': 38,
 'B73': 39,
 'B77': 40,
 'B78': 41,
 'B79': 42,
 'B80': 43,
 'B82 B84': 44,
 'B86': 45,
 'B94': 46,
 'B96 B98': 47,
 'C101': 48,
 'C103': 49,
 'C104': 50,
 'C106': 51,
 'C110': 52,
 'C111': 53,
 'C118': 54,
 'C123': 55,
 'C124': 56,
 'C125': 57,
 'C126': 58,
 'C128': 59,
 'C148': 60,
 'C2': 61,
 'C22 C26': 62,
 'C23 C25 C27': 63,
 'C30': 64,
 'C32': 65,
 'C45': 66,
 'C46': 67,
 'C47': 68,
 'C49': 69,
 'C50': 70,
 'C52': 71,
 'C54': 72,
 'C62 C64': 73,
 'C65': 74,
 'C68': 75,
 'C7': 76,
 'C70': 77,
 'C78': 78,
 'C82':

In [95]:
df['Cabin_target_encoded'] = df.Cabin.map(Encoded_labels_rank)
df.head(5)

PassengerId  Survived Pclass  \
0            1         0  Third   
1            2         1  First   
2            3         1  Third   
3            4         1  First   
4            5         0  Third   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Ordninal_Pclass  \
0      0         A/5 21171   7.2500   NaN        S                3   
1      0          PC 17599  71.2833   C85        C                1   
2      0  STON/O2. 3101282   7.9250   NaN        S                3   
3      0            113803  53.1000  C123        S                1   
4      0            373450   8.0500   NaN        S                3   

   Frequency_Encoded_Pclass  Encoded_cabin Cabin_first_letter  \
0                       491            NaN                  n   
1                       216            1.0                  C   
2                       491            NaN                  n   
3                       216            2.0                  C   
4                       491            NaN                  n   

   Cabin_target_encoded  
0                   NaN  
1                  81.0  
2                   NaN  
3                  55.0  
4                   NaN

In [106]:
Encoded_pClass = df.groupby(['Pclass'])['Survived'].mean().sort_values().index
Encoded_pClass

Index(['Third', 'Second', 'First'], dtype='object', name='Pclass')

In [107]:
Encoded_Pclass_dict = {k:i for i,k in enumerate(Encoded_pClass,0)}
Encoded_Pclass_dict

{'Third': 0, 'Second': 1, 'First': 2}

In [108]:
df['Target_encoded_Pclass'] = df.Pclass.map(Encoded_Pclass_dict)
df.head(5)

PassengerId  Survived Pclass  \
0            1         0  Third   
1            2         1  First   
2            3         1  Third   
3            4         1  First   
4            5         0  Third   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Ordninal_Pclass  \
0      0         A/5 21171   7.2500   NaN        S                3   
1      0          PC 17599  71.2833   C85        C                1   
2      0  STON/O2. 3101282   7.9250   NaN        S                3   
3      0            113803  53.1000  C123        S                1   
4      0            373450   8.0500   NaN        S                3   

   Frequency_Encoded_Pclass  Encoded_cabin Cabin_first_letter  \
0                       491            NaN                  n   
1                       216            1.0                  C   
2                       491            NaN                  n   
3                       216            2.0                  C   
4                       491            NaN                  n   

   Cabin_target_encoded  Target_encoded_Pclass  
0                   NaN                      0  
1                  81.0                      2  
2                   NaN                      0  
3                  55.0                      2  
4                   NaN                      0

### 5. Mean Encoding

In [111]:
Encoded_Mean_Dict = df.groupby(['Cabin'])['Survived'].mean().to_dict()
Encoded_Mean_Dict

{'A10': 0.0,
 'A14': 0.0,
 'A16': 1.0,
 'A19': 0.0,
 'A20': 1.0,
 'A23': 1.0,
 'A24': 0.0,
 'A26': 1.0,
 'A31': 1.0,
 'A32': 0.0,
 'A34': 1.0,
 'A36': 0.0,
 'A5': 0.0,
 'A6': 1.0,
 'A7': 0.0,
 'B101': 1.0,
 'B102': 0.0,
 'B18': 1.0,
 'B19': 0.0,
 'B20': 1.0,
 'B22': 0.5,
 'B28': 1.0,
 'B3': 1.0,
 'B30': 0.0,
 'B35': 1.0,
 'B37': 0.0,
 'B38': 0.0,
 'B39': 1.0,
 'B4': 1.0,
 'B41': 1.0,
 'B42': 1.0,
 'B49': 1.0,
 'B5': 1.0,
 'B50': 1.0,
 'B51 B53 B55': 0.5,
 'B57 B59 B63 B66': 1.0,
 'B58 B60': 0.5,
 'B69': 1.0,
 'B71': 0.0,
 'B73': 1.0,
 'B77': 1.0,
 'B78': 1.0,
 'B79': 1.0,
 'B80': 1.0,
 'B82 B84': 0.0,
 'B86': 0.0,
 'B94': 0.0,
 'B96 B98': 1.0,
 'C101': 1.0,
 'C103': 1.0,
 'C104': 1.0,
 'C106': 1.0,
 'C110': 0.0,
 'C111': 0.0,
 'C118': 0.0,
 'C123': 0.5,
 'C124': 0.0,
 'C125': 1.0,
 'C126': 1.0,
 'C128': 0.0,
 'C148': 1.0,
 'C2': 0.5,
 'C22 C26': 0.3333333333333333,
 'C23 C25 C27': 0.5,
 'C30': 0.0,
 'C32': 1.0,
 'C45': 1.0,
 'C46': 0.0,
 'C47': 1.0,
 'C49': 0.0,
 'C50': 1.0,
 'C52': 1.

In [112]:
df['Cabin_Mean_encoded'] = df.Cabin.map(Encoded_Mean_Dict)
df.head(5)

PassengerId  Survived Pclass  \
0            1         0  Third   
1            2         1  First   
2            3         1  Third   
3            4         1  First   
4            5         0  Third   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Ordninal_Pclass  \
0      0         A/5 21171   7.2500   NaN        S                3   
1      0          PC 17599  71.2833   C85        C                1   
2      0  STON/O2. 3101282   7.9250   NaN        S                3   
3      0            113803  53.1000  C123        S                1   
4      0            373450   8.0500   NaN        S                3   

   Frequency_Encoded_Pclass  Encoded_cabin Cabin_first_letter  \
0                       491            NaN                  n   
1                       216            1.0                  C   
2                       491            NaN                  n   
3                       216            2.0                  C   
4                       491            NaN                  n   

   Cabin_target_encoded  Target_encoded_Pclass  Cabin_Mean_encoded  
0                   NaN                      0                 NaN  
1                  81.0                      2                 1.0  
2                   NaN                      0                 NaN  
3                  55.0                      2                 0.5  
4                   NaN                      0                 NaN

### 5. Probability encoding --> It is mostly for the binary classification

In [119]:
Probability_survived = df.groupby(['Cabin_first_letter'])['Survived'].mean()
Probability_survived = pd.DataFrame(Probability_survived)
Probability_survived['Died'] = 1- Probability_survived
Probability_survived['Prob_ratio'] = Probability_survived['Survived']/Probability_survived['Died']
Probability_survived

Survived      Died  Prob_ratio
Cabin_first_letter                                
A                   0.466667  0.533333    0.875000
B                   0.744681  0.255319    2.916667
C                   0.593220  0.406780    1.458333
D                   0.757576  0.242424    3.125000
E                   0.750000  0.250000    3.000000
F                   0.615385  0.384615    1.600000
G                   0.500000  0.500000    1.000000
T                   0.000000  1.000000    0.000000
n                   0.299854  0.700146    0.428274

In [120]:
Encode_Prob_ratio_cabin = Probability_survived['Prob_ratio'].to_dict()
Encode_Prob_ratio_cabin

{'A': 0.875,
 'B': 2.916666666666666,
 'C': 1.4583333333333333,
 'D': 3.125,
 'E': 3.0,
 'F': 1.6000000000000003,
 'G': 1.0,
 'T': 0.0,
 'n': 0.42827442827442824}

In [121]:
df['Encoded_Prob_ratio_cabin'] = df.Cabin_first_letter.map(Encode_Prob_ratio_cabin)
df

PassengerId  Survived  Pclass  \
0              1         0   Third   
1              2         1   First   
2              3         1   Third   
3              4         1   First   
4              5         0   Third   
..           ...       ...     ...   
886          887         0  Second   
887          888         1   First   
888          889         0   Third   
889          890         1   First   
890          891         0   Third   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  Ordninal_Pclass  \
0        0         A/5 21171   7.2500   NaN        S                3   
1        0          PC 17599  71.2833   C85        C                1   
2        0  STON/O2. 3101282   7.9250   NaN        S                3   
3        0            113803  53.1000  C123        S                1   
4        0            373450   8.0500   NaN        S                3   
..     ...               ...      ...   ...      ...              ...   
886      0            211536  13.0000   NaN        S                2   
887      0            112053  30.0000   B42        S                1   
888      2        W./C. 6607  23.4500   NaN        S                3   
889      0            111369  30.0000  C148        C                1   
890      0            370376   7.7500   NaN        Q                3   

     Frequency_Encoded_Pclass  Encoded_cabin Cabin_first_letter  \
0                         491            NaN                  n   
1                         216            1.0                  C   
2                         491            NaN                  n   
3                         216            2.0                  C   
4                         491            NaN                  n   
..                        ...            ...                ...   
886                       184            NaN                  n   
887                       216            1.0                  B   
888                       491            NaN                  n   
889                       216            1.0                  C   
890                       491            NaN                  n   

     Cabin_target_encoded  Target_encoded_Pclass  Cabin_Mean_encoded  \
0                     NaN                      0                 NaN   
1                    81.0                      2                 1.0   
2                     NaN                      0                 NaN   
3                    55.0                      2                 0.5   
4                     NaN                      0                 NaN   
..                    ...                    ...                 ...   
886                   NaN                      1                 NaN   
887                  30.0                      2                 1.0   
888                   NaN                      0                 NaN   
889                  60.0                      2                 1.0   
890                   NaN                      0                 NaN   

     Encoded_Prob_ratio_cabin  
0                   